app/i18n/translator.html

```diff
@@ -128,111 +169,269 @@
         return el;
       }
 
-      function prepareTranslatedFile() {
+      function exportTranslatedFile() {
+        saveTranslation();
         const translated = state.translated;
-        let data = Object.keys(translated).sort().map(k => `  "${k}": ${JSON.stringify(translated[k])}`).join(",\n");
+        let data = Object.keys(translated)
+          .sort()
+          .map((k) => `  "${k}": ${JSON.stringify(translated[k])}`)
+          .join(",\n");
         let json = `{\n${data}\n}\n`;
-        const file = new Blob([json], {type: "application/octet-stream"});
+        const file = new Blob([json], { type: "application/octet-stream" });
         const url = URL.createObjectURL(file);
 
-        const downBtn = elID("down_translation_btn");
+        const downBtn = elID("export_translation_btn");
         downBtn.href = url;
         downBtn.download = "custom-translation.json";
       }
 
+      function isEntryDeleted(key) {
+        return state.deletedEntries.indexOf(key) !== -1;
+      }
+
       function saveTranslation() {
         if (state.selectedKey) {
+          const isDeleted = isEntryDeleted(state.selectedKey);
           // Save last translation.
-          state.translated[state.selectedKey] = elID("translated_str").value;
+          if (!isDeleted) {
+            const newValue = elID("translated_str").value;
+            state.translated[state.selectedKey] = newValue;
+
+            // update untranslated list if necessary
+            const el = elID(`li-untranslated-${state.selectedKey}`);
+            if (newValue !== "" && el) {
+              el.remove();
+            } else if (newValue === "" && !el) {
+              const stringList = elID("string_list_untranslated");
+              stringList.innerHTML = "";
+            }
```

Comment:
I have not tested this code path and wonder what it does?

Reads like "if new translated value is empty AND the element for selected key not found in the untranslated list, then... clear the entire untranslated list"?

Reply:
You read the code well.
If the user, for some reason, clears a previously translated translation, the item becomes untranslated. It should be added back to the untranslated list and be rerendered by clearing the related DOM. Another solution would be just adding back only the certain item. Since it could be a rare event, I think it's not too expensive to rerender the entire untranslated list with the proper order of the original list.

Reply:
It makes sense now, thanks.